File: leetcode/3330_find_the_original_typed_string_i.cpp
#include <iostream>
#include <stack>
#include <string>

using namespace std;
class Solution {
  public:
    static int possibleStringCount(const string &word) {
        // return two_ptr_stack(word);
        return one_loop(word);
    }
    // clang-format off
    static int two_ptr_stack(const string word) {
        stack<char> sk;
        int ans{};
        for (const char ch : word)
            if      (sk.empty())     sk.push(ch);
            else if (sk.top() == ch) ++ans;
            else                     sk.push(ch);
        return ans + 1;
    }
    static int one_loop(const string word) {
        int ans{};
        for (int i{}; i < word.length() - 1; ++i)
            ans += word[i] == word[i + 1] ? 1 : 0;
        return ans + 1;
    }
};

int main() {
    Solution s = Solution{};
    cout << "ans: " << Solution::possibleStringCount("abbcccc") << '\n'; // 5
    cout << "ans: " << Solution::possibleStringCount("abcd") << '\n';    // 1
    cout << "ans: " << Solution::possibleStringCount("aaaa") << '\n';    // 4
    cout << "ans: " << Solution::possibleStringCount("ere") << '\n';     // 1
    cout << "ans: " << s.possibleStringCount("asd") << '\n';             // 1
}

File: leetcode/3330_find_the_original_typed_string_i.rs
struct Solution;
impl Solution {
    fn possible_string_count(word: String) -> i32 {
        word.chars()
            .collect::<Vec<_>>()
            .windows(2)
            // .inspect(|x| println!("{x:?}"))
            .fold(0, |ans, win| if win[0] == win[1] { ans + 1 } else { ans })
            + 1
    }
}

fn main() {
    // $cargo eval --test -- path/to/this/file
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn testcase() {
        // 有件好fun 嘅事，下面個testcase 打少咗個c，搞咗半個鐘先發現...
        assert_eq!(Solution::possible_string_count("abbcccc".to_string()), 5);
        assert_eq!(Solution::possible_string_count("abcd".to_string()), 1);
        assert_eq!(Solution::possible_string_count("aaaa".to_string()), 4);
        assert_eq!(Solution::possible_string_count("ere".to_string()), 1);
        assert_eq!(Solution::possible_string_count("asd".to_string()), 1);
    }
}
